### ISSUE TITLE
Analog signal is not filtered

### ISSUE THREAD

**Issue by shashwatsridhar** [spyking-circus/spyking-circus#119]
Hey,

I am trying to run Spyking Circus on data from [this](https://web.gin.g-node.org/INT/multielectrode_grasp) repository. I don't think my data is being filtered before the sorting, for some reason, and I can't figure out why exactly. I have converted the original data to the .npy format. To make it easier for you to reproduce the issue, I have uploaded a subset of the dataset (96 channels x 1000000 data points) [here](https://fz-juelich.sciebo.de/s/EJ9d6XQ3CctaU2W) (about 360mb of raw data). The link also contains the probefile and params file I used.

I use spyking-circus-launcher to perform the preview sorting. The resulting analog signal that you can see in the Phy GUI does not look filtered on any of the electrodes. 

![Screenshot at 2019-04-01 20-21-00](https://user-images.githubusercontent.com/16129966/55350378-42a1de00-54bc-11e9-984f-560e0f7c4b9b.png)

The output in the console of the the spyking-circus-launcher GUI is as follows:

 ```
Starting spyking circus at Mon Apr  1 20:21:33 2019.

Command line call:
spyking-circus /home/shashwat/Work/temp/data/output.npy --preview --cpu 4


-------------------------  Informations  -------------------------
| Preview mode, showing only seconds [0-1] of the recording
------------------------------------------------------------------

##################################################################
#####           Welcome to the SpyKING CIRCUS (0.8.0)        #####
#####                                                        #####
#####              Written by P.Yger and O.Marre             #####
##################################################################


File          : /home/shashwat/Work/temp/data/output.npy
Steps         : preview mode
Number of CPU : 4/4
Parallel HDF5 : False
Shared memory : True
Hostfile      : /home/shashwat/spyking-circus/circus.hosts

##################################################################


-------------------------  Informations  -------------------------
| Number of recorded channels : 96
| Number of analyzed channels : 96
| File format                 : NUMPY
| Data type                   : float32
| Sampling rate               : 30 kHz
| Duration of the recording   : 0 min 33 s 333 ms
| Width of the templates      : 3 ms
| Spatial radius considered   : 100 um
| Threshold crossing          : negative
------------------------------------------------------------------
Filtering the signal with a Butterworth filter in (250, 5000) Hz
100%||[00:00<00:00,  3.01it/s]                   |[00:00<?, ?it/s]
Analyzing data to get whitening matrices and thresholds...
Found 0.130473s without spikes for whitening matrices...
Because of whitening, need to recompute the thresholds...
Searching spikes to construct the PCA basis...
100%||[00:00<00:00,  4.99it/s]<00:00,  4.92it/s] |[00:00<?, ?it/s]
Found 123 waveforms over 76800 requested
-------------------------  Informations  -------------------------
| A basis with 5 dimensions has been built
------------------------------------------------------------------

Process exited normally
```

I tried the same procedure by converting my data to raw_binary files saved in the .dat format (the link above also contains this file). The filtering results were different but still not correct. The data doesn't look like it's being properly filtered. Here's a screenshot (of the result view in this case).

![dat_result](https://user-images.githubusercontent.com/16129966/55368933-9d572c00-54f4-11e9-8cac-f3e038c238df.png)

Any clue what might be wrong?

Thanks! :)

EDIT: Forgot to mention that I am running Spyking Circus v0.8.0 (local install using `python setup.py develop`) on an Ubuntu 18.04 machine. numpy version 1.16.2.

**Comment by yger** [comment]
The data in the numpy format do not make sense to me. I don't really know
how they have been saved, but they are not the same as the one in the raw
binary file (see figure). If I'm processing the raw binary file, assuming
data are in float32, I do indeed see what I think are spikes on some
channels, and the filtering is working fine. So I think the code is working
fine. But some channels have a weird activity (see attached plot). Are you
sure all recording channels are extracellular recordings? Or are you sure
of the way they have been exported?



Le mar. 2 avr. 2019 à 03:15, Shashwat Sridhar <notifications@github.com> a
écrit :

**Comment by shashwatsridhar** [comment]
Hey Pierre,

Thanks for the quick reply, I spent a lot of time in the last few days verifying the scripts that I use to load and convert the data to npy/dat. I am fairly certain the data in the files I sent you is fine. In any case, here is the script I use to convert my data to npy/dat.

```python
import numpy as np
import os
import sys
from neo.io.blackrockio_v4 import BlackrockIO

if __name__ == '__main__':

    try:
        workingfile = sys.argv[1]
        workingpath = sys.argv[2]
    except IndexError:
        print("Arguments required by converting file: File name and Folder path!", file=sys.stderr)
        sys.exit(0)

    session_path = os.path.join(workingpath, workingfile)

    # load and convert data
    session = BlackrockIO(session_path)
    print("\nLoading raw analog signals for conversion", file=sys.stdout)
    channel_list = range(1, 97)
    signal_list = []
    for c in channel_list:
        raw = session.read_block(channels=c, units=None, load_events=False,
                                 load_waveforms=False, nsx_to_load=6)
        anasig = raw.segments[0].analogsignals[0].magnitude.T[0].astype(np.float32)
        signal_list.append(anasig)

    anasigs = np.array(signal_list).astype(np.float32)
    np.save(session_path, anasigs)

    anasigs = anasigs.T.flatten()
    anasigs = bytearray(anasigs)
    with open(session_path + ".dat", "wb+") as outfile:
        outfile.write(anasigs)

```

This saves both the raw_binay (.dat) and the numpy file.

I have now uploaded a file called `plot_pg.py` into the same folder (link again [here](https://fz-juelich.sciebo.de/s/EJ9d6XQ3CctaU2W)). It plots the first few channels of the dataset using pyqtgraph. More importantly, it uses the same data loaded routines used in spyking circus (`np.memmap` for npy files, `np.lib.format.open_memmap` for .dat files). To run it, you need a simple conda/pip environment with python3 and these four packages - `numpy pyqtgraph scipy docopt`, The instructions to use the file are given in its docstring, but here's an example if you want to run it using the data in the shared directory.

For the numpy file:
`python plot_pg.py npy attempt_float 10 60000`

For the raw_binary file:
`python plot_pg.py dat trimmed_data 10 60000 96 np.float32`

The script plots the raw signal (in red) and the filtered signal (in green) on the same plot for each channel. The filtered signals look fine in these plots, for both - .npy and .dat. I can't figure out what might be going wrong when I load the same data in spyking circus.

One guess I have is the data type I am specifying in the params file. If I have saved the data in the np.float32 format, is there anyway I can specify this in the params file. Is the float32 the same as np.float32?

**Comment by yger** [comment]
I think I found your problem: by looking at your script, I see that you are
exporting only 96 channels in the raw binary. So why are you stating, in
the probe file, that you have 132 channels ? If you set total_nb_channels
to 96, the data are making sense !

Best

Pierre

Le ven. 5 avr. 2019 à 17:26, Shashwat Sridhar <notifications@github.com> a
écrit :

**Comment by shashwatsridhar** [comment]
Thank you! That solved the problem :)
I had set the total number of channels to 132 when I was setting up the probe file to mimic all the electrodes of the Utah Array. I thought spyking circus would automatically "understand" that I only wanted the first 96 channels to be analyzed if I specified this is the first channel_group. But anyway, now I have much better sorting results :)

Thanks, again!
